### PR TITLE
[PF-560] Add Harness desktop workspace policy evaluator

### PR DESCRIPTION
**Pull request by mbenhamd**
## Summary
- Add a public Harness v1 workspace policy evaluator for desktop hosts and constrained executors.
- Resolve file paths and command cwd/root context against declared workspace roots.
- Enforce traversal denial, read-only root mutation denial, deterministic deny > ask > allow precedence, cross-root rename safety, command cwd/root scoping, and network/MCP selector matching.
- Export the evaluator/types from @mastra/core/harness/v1 and add a patch changeset with a public usage example.

## Scope
- PF-560 child of PF-551.
- Local core contract only: no server routes, no storage schema, no OS sandboxing, no production actions.
- Open PR overlap check before create: none open in mbenhamd/mastra.
- Rebased onto latest origin/main before opening: d176682677e40873e7a415136960534a56cede0a.

## Validation
- pnpm install --frozen-lockfile (needed because this worktree had no node_modules; completed with existing ignored-build-script warnings)
- pnpm build:core (needed to populate @internal/test-utils setup after initial focused test could not resolve it)
- pnpm --filter ./packages/core test:unit src/harness/v1/workspace-policy.test.ts
- pnpm --filter ./packages/core check
- pnpm prettier --check packages/core/src/harness/v1/workspace-policy.ts packages/core/src/harness/v1/workspace-policy.test.ts packages/core/src/harness/v1/index.ts .changeset/upset-corners-join.md
- git diff --check
- Local Codex review pass 1: fixed target-root rename, nested read-only root, and command cwd findings
- Local Codex review pass 2: fixed command rootId and network/MCP selector API findings
- coderabbit review --plain: fixed changeset description finding; rerun completed with no findings

## Follow-up
- Later PF-551 slices should wire this evaluator into concrete desktop host file/command/network/MCP execution surfaces, audit journal, and restore flows.

## ELI5

This PR adds a simple safety checker that, before desktop or constrained executors touch files, run commands, use the network, or call AI tools, verifies those actions against workspace rules (with "deny" taking priority) and returns resolved paths plus rule reasons for approvals and audit.

## Overview

Adds a public Harness v1 workspace policy evaluator to `@mastra/core` that:
- Resolves action paths against declared workspace roots (POSIX and Windows-aware) and selects the most-specific root.
- Enforces no-traversal outside configured roots (including symlink escape detection), denies mutating ops on read-only roots, and enforces cross-root rename safety.
- Scopes selectors to action kinds (file, command, network, MCP) and enforces command cwd/root scoping.
- Produces deterministic decisions with precedence deny > ask > allow, falling back to policy.defaultDecision (defaults to ask).
- Returns enriched evaluation results including resolved path / toPath / cwd / root and matched rule reasons for approval screens and audit logs.

Scope: core contract only (no server routes, storage schema, OS sandboxing, or runtime wiring).

## Changes

- packages/core/src/harness/v1/workspace-policy.ts
  - New evaluator and helpers:
    - resolveWorkspacePath(roots, inputPath, rootId?) — Windows/POSIX handling, normalization, containment checks, and most-specific-root selection.
    - evaluateWorkspacePolicy(policy, action) — resolves path/toPath/cwd, enforces writable-root restrictions for mutating ops, matches rules by action kind and selectors (file root, command root/cwd, network host/port/protocol, MCP server/tool), aggregates matched reasons, applies deterministic decision precedence, and handles rename/source-target validation and symlink escapes.
  - New exported types/interfaces: WorkspacePolicy, WorkspacePolicyRule, WorkspacePolicyAction (and per-kind variants), WorkspaceResolvedPath, WorkspacePolicyEvaluation, WorkspacePolicyMatchedRule, WorkspaceRootDescriptor, WorkspaceFileOperation, WorkspacePolicyActionKind, etc.

- packages/core/src/harness/v1/index.ts
  - Re-exports evaluateWorkspacePolicy, resolveWorkspacePath, and workspace-policy types from ./workspace-policy.

- packages/core/src/harness/v1/workspace-policy.test.ts
  - New Vitest suite covering:
    - Most-specific-root selection (including Windows absolute paths).
    - Path traversal prevention (POSIX and Windows separators), segment-aware containment, and symlink escape denial.
    - Preservation of literal backslashes for POSIX roots and mixed-separator edge cases.
    - Read-only root denial for mutating ops and writable defaulting behavior.
    - Rename safety across roots, rule application to source vs. target, and missing-target checks.
    - Rule precedence (deny > ask > allow), ask vs allow resolution, and defaultDecision fallback.
    - Selector scoping per action kind and negative cross-kind cases.
    - Command cwd containment and rootId scoping/validation (including workspace.root_not_found).

- .changeset/upset-corners-join.md
  - Patch release note documenting the new workspace policy evaluation capability and a usage example showing evaluateWorkspacePolicy and throwing on non-allow decisions (including reasons).

- packages/core/src/harness/v1/index.ts
  - Re-exports added for evaluateWorkspacePolicy, resolveWorkspacePath, and related types.

## Validation & Reviews

- Local validation performed: pnpm install --frozen-lockfile; pnpm build:core; unit tests for workspace-policy.test.ts; pnpm --filter ./packages/core check; prettier checks on specified files; git diff --check.
- Two Codex review passes and a coderabbit review completed with fixes applied.
- Branch rebased onto origin/main commit d176682677e40873e7a415136960534a56cede0a. PR PF-560 is a child of PF-551.

## Follow-up Work

Wire the evaluator into desktop host file/command/network/MCP execution surfaces, audit journal, and restore flows in later PF-551 slices.

## Notes for reviewers

- Implementation intentionally focuses on core policy evaluation (no runtime integrations or sandboxing).
- Tests include mixed path-separator and Windows/POSIX edge cases added after review.

[Review Change Stack](https://app.coderabbit.ai/change-stack/mbenhamd/mastra/pull/152?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: .changeset/upset-corners-join.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+---
+'@mastra/core': patch
+---
+
+Added Harness desktop workspace policy evaluation for desktop hosts and other constrained executors.
+
+Apps can now enforce file, command, network, and MCP access rules before starting work. The evaluator resolves file paths and command working directories against declared workspace roots, rejects path traversal and read-only root writes, applies deterministic `deny > ask > allow` precedence, and returns the matched rules and resolved paths for audit or approval UI.
```
[the remark this answers is not in the document]
Comment (mbenhamd) [reply .changeset/upset-corners-join.md]: Agent-authored with user authorization: addressed in commit 911d38e68f. The changeset now uses short outcome-focused bullets and expands Model Context Protocol (MCP). Validated with prettier --check, git diff --check, workspace-policy focused tests, and core typecheck.

File: .changeset/upset-corners-join.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    roots: [{ id: 'project', path: '/workspace/project', writable: true }],
+    defaultDecision: 'deny',
+    rules: [{ kind: 'file', rootId: 'project', operation: 'write', decision: 'ask' }],
+  },
+  { kind: 'file', operation: 'write', path: 'src/index.ts', rootId: 'project' },
+);
+
+if (result.decision === 'deny') {
+  throw new Error(result.reasons.join(', '));
+}
```
[the remark this answers is not in the document]
Comment (mbenhamd) [reply .changeset/upset-corners-join.md:28]: Agent-authored with user authorization: addressed in commit 911d38e68f. The public example now blocks every non-allow decision, including ask, before policy-gated execution proceeds. Validated with prettier --check, git diff --check, workspace-policy focused tests, and core typecheck.

File: packages/core/src/harness/v1/workspace-policy.ts [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+      return {
+        decision: 'deny',
+        reasons: [resolved.reason],
+        matchedRules: [],
+      };
+    }
+    resolvedPath = resolved.path;
+    resolvedToPath = resolved.toPath;
+    if (isMutatingFileOperation(action.operation)) {
+      const readOnlyRoot = [resolvedPath.root, resolvedToPath?.root].find(root => root?.writable === false);
```
[the remark this answers is not in the document]
Comment (mbenhamd) [reply packages/core/src/harness/v1/workspace-policy.ts:121]: Agent-authored with user authorization: addressed in commit 911d38e68f. WorkspaceRootDescriptor now documents that omitted writable defaults to writable, and tests cover write plus rename behavior with omitted writable roots. Validated with prettier --check, git diff --check, workspace-policy focused tests, and core typecheck.

File: packages/core/src/harness/v1/workspace-policy.ts [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+    .sort((left, right) => right.root.path.length - left.root.path.length);
+
+  return resolved[0] ?? null;
+}
+
+function resolveAgainstRoot(root: WorkspaceRootDescriptor, inputPath: string): WorkspaceResolvedPath | null {
+  if (!isUsablePath(root.id) || !isUsablePath(root.path)) return null;
+
+  const flavor = pathFlavorFor(root.path, inputPath);
+  const pathApi = flavor === 'win32' ? path.win32 : path.posix;
```
[the remark this answers is not in the document]
Comment (mbenhamd) [reply packages/core/src/harness/v1/workspace-policy.ts:257]: Agent-authored with user authorization: addressed in commits 911d38e68f and f271fbd886. Tests now cover Windows absolute path resolution, Windows-style traversal denial, and POSIX roots treating backslashes as literal path characters. Validated with workspace-policy focused tests, prettier --check, and git diff --check.

File: packages/core/src/harness/v1/workspace-policy.ts [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+
+function resolveWorkspaceActionPath(
+  roots: readonly WorkspaceRootDescriptor[],
+  inputPath: string,
+  rootId?: string,
+): WorkspaceResolvedPath | null {
+  const requestedPath = resolveWorkspacePath(roots, inputPath, rootId);
+  if (!requestedPath) return null;
+
+  return resolveWorkspacePath(roots, requestedPath.normalizedPath) ?? requestedPath;
```
[the remark this answers is not in the document]
Comment (mbenhamd) [reply packages/core/src/harness/v1/workspace-policy.ts]: Agent-authored with user authorization: fixed in `8884bcba24`. `resolveWorkspaceActionPath()` now preserves an explicitly selected root when canonical re-resolution finds another root with the same normalized filesystem path, so duplicate-path roots cannot swap read-only/root-scoped policy context. It still allows more-specific nested roots to take precedence. Added regression coverage for duplicate-path writable/read-only roots. Validation: focused workspace-policy tests passed (20 tests), `pnpm --filter ./packages/core check` passed, `git diff --check` passed, and CodeRabbit CLI returned `findings: 0`. Resolving as addressed.

File: packages/core/src/harness/v1/workspace-policy.ts [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+  | { status: 'allowed'; path: WorkspaceResolvedPath; toPath?: WorkspaceResolvedPath }
+  | { status: 'denied'; reason: string } {
+  const resolvedPath = resolveWorkspaceActionPath(roots, action.path, action.rootId);
+  if (!resolvedPath) {
+    return { status: 'denied', reason: 'workspace.path_outside_roots' };
+  }
+
+  if (action.toPath === undefined) {
+    return { status: 'allowed', path: resolvedPath };
+  }
```
[the remark this answers is not in the document]
Comment (mbenhamd) [reply packages/core/src/harness/v1/workspace-policy.ts:193]: Agent-authored with user authorization: fixed in `8884bcba24`. `resolveWorkspaceFileAction()` now rejects `operation: "rename"` without `toPath` with `workspace.target_path_required` instead of falling through to the default policy. Added a regression test for the malformed rename case. Validation: focused workspace-policy tests passed (20 tests), core typecheck passed, and CodeRabbit CLI returned `findings: 0`. Resolving as addressed.

File: packages/core/src/harness/v1/workspace-policy.ts [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+      rule.mcpToolName === undefined
+    );
+  }
+  if (rule.rootId !== undefined) return false;
+
+  if (action.kind === 'network') {
+    return (
+      matchesValue(rule.networkHost, action.host) &&
+      matchesValue(rule.networkPort, action.port) &&
+      matchesValue(rule.networkProtocol, action.protocol) &&
```
[the remark this answers is not in the document]
Comment (mbenhamd) [reply packages/core/src/harness/v1/workspace-policy.ts]: Agent-authored with user authorization: fixed in `8884bcba24`. Network protocol matching now normalizes selector and action values by lowercasing and stripping one trailing colon, so `https:` from `URL.protocol` matches policy rules written as `https`. Added regression coverage through the existing network selector test. Validation: focused workspace-policy tests passed (20 tests), core typecheck passed, and CodeRabbit CLI returned `findings: 0`. Resolving as addressed.

File: packages/core/src/harness/v1/workspace-policy.ts [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+
+  const normalizedPath = pathApi.isAbsolute(inputPath)
+    ? pathApi.normalize(inputPath)
+    : pathApi.resolve(normalizedRoot, inputPath);
+  const relativePath = pathApi.relative(normalizedRoot, normalizedPath);
+
+  if (relativePath === '') {
+    return { root: { ...root, path: normalizedRoot }, normalizedPath, relativePath: '.' };
+  }
+  if (isOutsideRelativePath(relativePath, pathApi)) return null;
```
[the remark this answers is not in the document]
Comment (mbenhamd) [reply packages/core/src/harness/v1/workspace-policy.ts:269]: Agent-authored with user authorization: fixed in `8884bcba24`. Path resolution now performs a best-effort realpath containment check when the workspace root and target or nearest existing parent exist on disk. Existing symlink paths that resolve outside the declared root are denied before rule matching. Added a regression test that creates a real symlink inside a temporary root pointing outside the root and verifies a write through it is denied. Validation: focused workspace-policy tests passed (20 tests), core typecheck passed, and CodeRabbit CLI returned `findings: 0`. Resolving as addressed.

File: packages/core/src/harness/v1/workspace-policy.ts [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+      rule.networkProtocol === undefined &&
+      rule.mcpServerId === undefined &&
+      rule.mcpToolName === undefined
+    );
+  }
+  if (rule.rootId !== undefined) return false;
+
+  if (action.kind === 'network') {
+    return (
+      matchesValue(rule.networkHost, action.host) &&
```
[the remark this answers is not in the document]
Comment (mbenhamd) [reply packages/core/src/harness/v1/workspace-policy.ts]: Agent-authored with user authorization: fixed in `8884bcba24`. Network host matching now lowercases selector and action values before comparing, so deny/ask rules configured for `api.example.test` match caller input such as `API.EXAMPLE.TEST`. Added regression coverage through the existing network selector test. Validation: focused workspace-policy tests passed (20 tests), core typecheck passed, and CodeRabbit CLI returned `findings: 0`. Resolving as addressed.

**Comment by mbenhamd** [comment]
Agent-authored with user authorization: addressed the CodeRabbit review-body nitpick about the `realpathIfExists` fallback in commit `a2315783b657c2e44661e45be77effac01656cf4` by documenting that missing or inaccessible paths fall back to the lexical containment checks. Local validation passed: `git diff --check`, `pnpm --filter ./packages/core exec eslint src/harness/v1/workspace-policy.ts`, and `pnpm --filter ./packages/core exec vitest run src/harness/v1/workspace-policy.test.ts`. Inline review threads are currently resolved; CodeRabbit status is still pending on the latest head, so I am not marking the PR merge-ready until it completes.
